Setup wizard: Done screen

johnwatkins0 opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Fixes #4707

This adds the final screen for the setup wizard. Uses are presented with an iframe showing an AMP page on their site in a mobile viewport, along with links to view the site or customize (in Reader mode). The "Next" button becomes "Finish" and links back to AMP settings.

<img width="1553" alt="Screen Shot 2020-06-29 at 11 04 03 PM" src="https://user-images.githubusercontent.com/9020968/86082261-e9c54a80-ba5c-11ea-88cc-2da2637b95dc.png">


## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Add done screen copy

johnwatkins0 committed:
Done screen styles

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/4707-done-screen

johnwatkins0 committed:
Setup wizard iframe src/links

johnwatkins0 committed:
Fix amp setup nav z-index

westonruter commented on assets/src/setup/pages/save/index.js:
```diff
@@ -41,16 +103,88 @@ export function Save() {
+				<Phone>
+					<iframe
+						className="done__preview-iframe"
+						sandbox="allow-scripts"
```
```suggestion
						sandbox="allow-scripts allow-forms allow-popups allow-presentation"
```

johnwatkins0 replied:
You got this in just ahead of my comment below. I'll try it in a little while.

johnwatkins0 commented:
@westonruter This PR is still WIP, but I wanted to flag for you that I made an attempt to hide the admin bar in the iframe using the `sandbox` attribute. I expected it to work, but for some reason it didn't. I could sense that figuring this out was potentially going to take a signicant amount of time, so I went to the fallback of setting a `amp-no-admin-bar` URL var and turning off the admin bar when it is set. 

This PR still has the iframe with the `sandbox` parameter. To replicate, remove the `show_admin_bar` filter I added in `includes/amp-helper-functions.php`. Then, in the wizard, select standard mode and navigate to the last screen, which has the iframe. Where we would expect the iframe to be hidden, it's still showing.

westonruter commented:
I may have been mistaken on the behavior of `sandbox` and whether it blocks cookies from being used then. Maybe it's just a matter of whether the iframed window can read cookies from the parent window or not.

westonruter commented:
I have another idea: provide a `name` to the `iframe`, like `amp-wizard-completion-preview`. Then, you could do something like this:

```php
add_action( 'wp_print_footer_scripts', function() {
	if ( ! amp_is_dev_mode() || ! is_admin_bar_showing() ) {
		return;
	}
	?>
	<script data-ampdevmode>
        document.addEventListener( 'DOMContentLoaded', () => {
            if ( document.body.classList.contains( 'admin-bar' ) ) {
                document.body.classList.remove( 'admin-bar' );
                document.documentElement.style.cssText += '; margin-top: 0 !important';
                document.getElementById( 'wpadminbar' ).remove();
            }
        } );
	</script>
	<?php
} );
```

But that would still result in a flash of the admin bar appearing momentarily. 

In any case, this is probably overkill. If the admin bar shows up upon navigation, that's not the end of the world 😄

johnwatkins0 commented:
This is very similar:

```PHP
/**
 * Hide admin bar if the window is inside the setup wizard iframe.
 */
add_action(
	'wp_print_footer_scripts',
	function() {
		if ( ! amp_is_dev_mode() || ! is_admin_bar_showing() ) {
			return;
		}
		?>
		<script data-ampdevmode>
			document.addEventListener( 'DOMContentLoaded', function() {
				if (!('name' in window) || 'amp-wizard-completion-preview' !== window.name) { // Detects whether the current window is in an iframe.
					return;
				}

				if ( document.body.classList.contains( 'admin-bar' ) ) {
					document.body.classList.remove( 'admin-bar' );
					document.documentElement.style.cssText += '; margin-top: 0 !important';
					document.getElementById( 'wpadminbar' ).remove();
				}
			});
		</script>
		<?php
	}
);
```

Apparently if a window is in an iframe, the iframe's `name` attribute is available as `window.name`. I've never come across this piece of information before (that I can recall). 

I don't think this is overkill because it's no more complicated and is maybe more reliable than the query var, and works across navigations, so I'm going to add it.

johnwatkins0 committed:
CSS polish and remove new onboarding feature flag

johnwatkins0 committed:
Testing and linting

johnwatkins0 committed:
Set max with on nav bar

johnwatkins0 committed:
Fix a failing PHPUnit test

johnwatkins0 committed:
Fix failing unit test in WP 4.9

johnwatkins0 commented:
Opening this for review after making a round of CSS updates not necessarily directly related to this task. I also **removed the feature flag** around the setup wizard, so anyone checking out this branch or `develop` after this branch is merged will have access to the setup wizard. No need for the `AMP_NEW_ONBOARDING` constant. This can easily be reversed if I've done it too soon.

johnwatkins0 commented:
@jwold The production build in the above comment should work in your local environment out of the box without having to mess with PHP constants.

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/4707-done-screen

jwold commented:
1. Can we make all the steps blue checkmarks as they get completed? https://d.pr/i/rDAmSj
2. I'm now wondering if we should have the gray outline: https://d.pr/i/ZMfD7t. It is feeling like a button. Curious for your thoughts! cc @amedina and @westonruter

johnwatkins0 committed:
Ensure SVG clip paths have unique IDs

johnwatkins0 commented:
> * Can we make all the steps blue checkmarks as they get completed? https://d.pr/i/rDAmSj
> * I'm now wondering if we should have the gray outline: https://d.pr/i/ZMfD7t. It is feeling like a button. Curious for your thoughts! cc @amedina and @westonruter

Now being handled in #4962

johnwatkins0 committed:
Force user to make technical question choice

johnwatkins0 committed:
Increase left padding on amp notice

johnwatkins0 committed:
Initialize mobile redirect option to true if wizard has never been completed

johnwatkins0 committed:
Padding adjustments

johnwatkins0 committed:
Remove mobile redirect icon

johnwatkins0 committed:
Prevent exiting while options are saving

westonruter commented on assets/src/setup/pages/save/index.js:
```diff
@@ -41,16 +106,84 @@ export function Save() {
+								{ __( 'Visit in Customizer', 'amp' ) }
+							</Button>
+							<Button
+								href={ addQueryArgs( PREVIEW_PERMALINK, { amp: 1 } ) }
```
See comment above about returning the permalink via the REST API.

westonruter commented on assets/src/setup/pages/save/index.js:
```diff
@@ -41,16 +106,84 @@ export function Save() {
+										CUSTOMIZER_LINK,
+										'legacy' === readerTheme
+											? { 'autofocus[panel]': 'amp_panel', url: PREVIEW_PERMALINK }
+											: { url: PREVIEW_PERMALINK },
```
The `amp` query var needs to be added when going to the non-Legacy Reader mode Customizer:

```suggestion
											: { url: PREVIEW_PERMALINK, amp: 1 },
```

johnwatkins0 replied:
I suppose this won't be needed with the preview permalink returned via the Options endpoint.

westonruter replied:
It will because this is an `amp` query var for `customize.php` not the `PREVIEW_PERMALINK`.

westonruter replied:
Granted, at the moment this does nothing. But it _will_.

johnwatkins0 replied:
OK, I hadn't picked up on the customizer link also needing the `amp` query var. Updated.

westonruter commented on assets/src/setup/pages/save/index.js:
```diff
@@ -41,16 +106,84 @@ export function Save() {
+				{ 'reader' !== themeSupport && (
+					<Button
+						isPrimary
+						href={
+							addQueryArgs(
+								PREVIEW_PERMALINK,
+								'transitional' === themeSupport
+									? { amp: 1 }
+									: {},
+							)
+						}
```
I suppose manipulating the query vars is needed because the output of `amp_admin_get_preview_permalink` may now be stale.

Here's an idea: can the REST API endpoint include the preview link URL? Perhaps a new endpoint? Or part of the options endpoint as a read-only value? Then when saving we will get the most up-to-date preview URL based on the settings. This will be very important when we allow sites to customize how the URLs are constructed for AMP.  See #2204.

johnwatkins0 replied:
I actually considered (1) including the preview link in the options response after saving, and (2) creating a new REST endpoint just for the preview link. 1 feels wrong because it's not an option, and 2 felt like overkill for a single field, but with your suggestion I'll go ahead with 1. It is sort of an option, just not one that persists in the DB.

westonruter commented on assets/src/setup/pages/save/index.js:
```diff
@@ -41,16 +106,84 @@ export function Save() {
+				<Phone>
+					<iframe
+						className="done__preview-iframe"
+						src={ addQueryArgs( PREVIEW_PERMALINK, 'standard' === themeSupport ? {} : { amp: 1 } ) }
```
See comment below about getting `PREVIEW_PERMALINK` from the REST API.

westonruter commented on includes/options/views/class-amp-setup-wizard-submenu-page.php:
```diff
@@ -148,6 +154,7 @@ public function enqueue_assets( $hook_suffix ) {
 			],
 			'OPTIONS_REST_ENDPOINT'              => rest_url( 'amp/v1/options' ),
 			'READER_THEMES_REST_ENDPOINT'        => rest_url( 'amp/v1/reader-themes' ),
+			'PREVIEW_PERMALINK'                  => amp_admin_get_preview_permalink(),
```
See comment above about how this should ideally be returned by the REST API after saving the mode.

westonruter replied:
```suggestion
```

johnwatkins0 committed:
Provide preview permalink over REST

johnwatkins0 committed:
Merge branch 'feature/4707-done-screen' into feature/4962-onboarding-wizard-ui-polish

johnwatkins0 committed:
Add amp query var to non-legacy reader customizer link

westonruter commented on includes/options/class-amp-options-rest-controller.php:
```diff
@@ -94,6 +101,11 @@ public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAna
 
 		$options = wp_array_slice_assoc( $options, array_keys( $properties ) );
+
+		// Add the preview permalink. The permalink can't be handled via AMP_Options_Manager::get_options because
+		// amp_admin_get_preview_permalink calls AMP_Options_Manager::get_options, leading to infinite recursion.
+		$options[ self::PREVIEW_PERMALINK ] = amp_admin_get_preview_permalink();
```
This is correct anyway, as the option is computed rather than something stored.

johnwatkins0 committed:
Remove obsolete PREVIEW_PERMALINK JS data

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Merge branch 'feature/4707-done-screen' into feature/4962-onboarding-wizard-ui-polish

westonruter commented on assets/src/setup/pages/save/style.css:
```diff
@@ -0,0 +1,35 @@
+
+.done__preview-iframe {
+	height: 610px;
+	width: 100%;
```
This is turning out to be extremely narrow, like 240px:

![image](https://user-images.githubusercontent.com/134745/86175827-d9eb4c00-bad8-11ea-86fd-b9a41525a9e6.png)

Therefore, I think the iframe needs to _not_ be floated to the right and instead appear _below_ the copy. The iframe should then have a `max-width:640px`.

westonruter replied:
Or perhaps a `max-width:480px`. Whatever the most common screen resolution is for mobile phones.

Example:

![image](https://user-images.githubusercontent.com/134745/86176300-947b4e80-bad9-11ea-8137-31c20f6e6cf1.png)

cc @jwold

johnwatkins0 replied:
I just set a fixed width of 400px on that column (approximately what it is on the design) and have it wrapping below a certain width) in a4003579f59bdb0cf23c4b8d3c9bf7f2e2834b29

<img width="1051" alt="Screen Shot 2020-06-30 at 4 29 49 PM" src="https://user-images.githubusercontent.com/9020968/86178978-04e09a80-baef-11ea-9d3a-3acb814de408.png">

jwold commented:
Let’s do it. Below is not as nice, but better. Can we just make it drop for smaller screens?

johnwatkins0 committed:
Minor CSS updates

johnwatkins0 committed:
Update snapshot

johnwatkins0 committed:
Visual improvements

jwold commented:
Looks good to me 
GitHub <notifications@github.com> wrote:
“@johnwatkins0 commented on this pull request.”


“In assets/src/setup/pages/save/style.css:”

> +	text-align: center;
+}
+
+.done__preview-container .phone {
+	max-height: 610px;
+	padding-bottom: 2rem;
+}
+
+.done__preview-container .phone::before {
+	height: 9px;
+	width: 80px;
+}
+
+.done__preview-iframe {
+	height: 610px;
+	width: 100%;
“I just set a fixed width of 400px on that column (approximately what it is on the design) and have it wrapping below a certain width).”

[Screen Shot 2020-06-30 at 4 29 49 PM]
“ —
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.”

johnwatkins0 committed:
Check current_user_can( 'install_themes' ) in can_install_themes check

westonruter committed:
Increase width of mobile preview iframe

westonruter committed:
Fix tests after 1743fde

westonruter committed:
Merge pull request #4966 from ampproject/feature/4962-onboarding-wizard-ui-polish

Onboarding wizard UI polish

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/4707-done-screen

* 'develop' of github.com:ampproject/amp-wp:
  Remove redundant units from CSS 0 lengths
  Unit test cleanup
  Test cleanup
  Update dependency source-map-loader to v1.0.1
  PHPCS fix
  Update configure link
  Add plugin activation notice

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into feature/4707-done-screen

* 'develop' of github.com:ampproject/amp-wp:
  Remove obsolete comment regarding replacing has_cap with is_user_enabled
  Remove unused get_dev_tools_user_access method
  Ensure Non-AMP link in admin bar is noamp
  Correct the order of arguments to assertEquals
  Reference the factory statically in tests
  Remove redundant type definition
  Use static closures
  Add tests
  Fix access to AMP Dev Tools when non-administrators granted access
  Use greater priority for map_meta_cap; add tests for has_cap and map_meta_cap
  Update WP-CLI validation run command to support (legacy) Reader mode
  Fix failing tests
  Show admin bar in Reader mode template
  Validate URLs in Reader mode if dev tools enabled
  Prevent checking for validation errors when activating plugins but dev tools disabled
  Eliminate dead code made obsolete by 3e2ba8a in #31190
  Introduce amp_validate capability and use it to restrict dev tools access

schlessera commented:
How about scaling the iframe content - http://jsfiddle.net/6th5107u/

westonruter commented:
> How about scaling the iframe content - http://jsfiddle.net/6th5107u/

That's a possibility, but the scaling causes visual regressions (e.g. text legibility):

![image](https://user-images.githubusercontent.com/134745/86204696-fc9f5400-bb1c-11ea-8c96-6b1eeb6062e9.png)

schlessera commented:
I'm getting an error when trying to complete the wizard:
![Image 2020-07-01 at 9 16 05 AM](https://user-images.githubusercontent.com/83631/86214577-a01f5180-bb7b-11ea-8fdd-196e9fca17dc.png)

I made the following choices:
- Developer or technically savvy
- Transitional (Previously selected)
- Astra theme
- Redirect mobile visitors to AMP

schlessera commented:
I think the bottom bar transparency looks off and is visually inconsistent.
![Image 2020-07-01 at 9 22 53 AM](https://user-images.githubusercontent.com/83631/86215170-8f231000-bb7c-11ea-9ab7-bdcad7180f1a.png)
![Image 2020-07-01 at 9 23 36 AM](https://user-images.githubusercontent.com/83631/86215178-921e0080-bb7c-11ea-848c-13f20810c83f.png)

I would personally prefer for this to stick more to the Material principle of everything having a consistent depth and shadows that represent that.
Here's the same bar without translucency and with a subtle shadow applied:
![Image 2020-07-01 at 9 22 02 AM](https://user-images.githubusercontent.com/83631/86215293-b679dd00-bb7c-11ea-96a6-ec70c937f0fe.png)
![Image 2020-07-01 at 9 22 08 AM](https://user-images.githubusercontent.com/83631/86215300-b7ab0a00-bb7c-11ea-96d2-38614bf7b6d6.png)

pierlon commented on assets/src/setup/components/nav/index.js:
```diff
@@ -64,11 +68,12 @@ export function Nav( { exitLink } ) {
 						onClick={ moveForward }
 					>
-						{ __( 'Next', 'amp' ) }
+						{ isLastPage ? __( 'Finish', 'amp' ) : __( 'Next', 'amp' ) }
```
I think we can remove the "next" arrow icon if it's the last page, otherwise it could imply there's an another page to naviagte to.

![image](https://user-images.githubusercontent.com/16200219/86224014-d78bff00-bb77-11ea-9df0-7f90dfcba9a6.png)

johnwatkins0 replied:
Good idea. Updated in 11adf2dcf

pierlon commented on assets/src/setup/components/nav/style.css:
```diff
@@ -10,17 +10,19 @@
 	align-items: center;
 	display: flex;
 	justify-content: space-between;
+	max-width: 1320px;
```
This makes the nav bar look misaligned from the setup wizard:

![image](https://user-images.githubusercontent.com/16200219/86224646-ab24b280-bb78-11ea-9338-cbe80ef8ba02.png)

johnwatkins0 replied:
Fixed in d570470a3

johnwatkins0 commented:
> I'm getting an error when trying to complete the wizard:
> ![Image 2020-07-01 at 9 16 05 AM](https://user-images.githubusercontent.com/83631/86214577-a01f5180-bb7b-11ea-8fdd-196e9fca17dc.png)

@schlessera This may be because you still have "classic" set as the reader theme, which won't be accepted by the `validate_callback` now that we've switched that slug to `legacy`. I think if you were to select reader mode and choose a reader theme from the list (you would even be able to go back and re-select transitional after choosing a theme) this issue would go away.

johnwatkins0 commented:
@jwold What do you think about @schlessera's suggestion on the bottom nav bar two comments up?

pierlon commented on assets/src/setup/pages/save/index.js:
```diff
@@ -2,26 +2,109 @@
 /**
  * Internal dependencies
  */
+import { Button } from '@wordpress/components';
```
WordPress dependency listed under internal.

pierlon commented on assets/src/setup/pages/save/index.js:
```diff
@@ -2,26 +2,109 @@
+	 */
+	useEffect( () => {
+		const timeout = setTimeout( () => {
+			setShouldShowSaving( false );
```
"Finish" button should be disabled when saving. Currently, it's enabled:

![image](https://user-images.githubusercontent.com/16200219/86257406-02db1200-bba9-11ea-99b1-1f9710d04a51.png)

johnwatkins0 replied:
Done (in a less hacky way than my previous attempt) in 48653bbd1

pierlon commented on assets/src/setup/pages/save/index.js:
```diff
@@ -31,26 +114,85 @@ export function Save() {
+			</div>
+			<div className="done__preview-container">
+				<AMPInfo icon={ ( props ) => <IconMobile { ...props } /> }>
+					{ __( 'Live view of your site', 'amp' ) }
```
Should we explicitly state that the preview is an AMP page? Otherwise, it may confuse users as to which version they are looking at (especially when in reader/transitional mode):

![image](https://user-images.githubusercontent.com/16200219/86258099-f4412a80-bba9-11ea-835e-8833800ceee0.png)


```suggestion
					{ __( 'Live view of your AMP site', 'amp' ) }
```

johnwatkins0 replied:
Done in 05ae9e993f8a3551eb55cf696d839022754ed5d9

jwold commented:
> @jwold What do you think about @schlessera's suggestion on the bottom nav bar two comments up?

Let's go with @schlessera's change. I think it will be an improvement. cc @westonruter and @amedina

pierlon commented on assets/src/setup/pages/save/index.js:
```diff
@@ -31,26 +114,85 @@ export function Save() {
+					{ __( 'Live view of your site', 'amp' ) }
+				</AMPInfo>
+				<Phone>
+					<iframe
```
Should the admin bar be shown when previewing the page with the reader mode template?

![image](https://user-images.githubusercontent.com/16200219/86258467-631e8380-bbaa-11ea-91ee-9a87a1a62cae.png)

johnwatkins0 replied:
I should defer to @westonruter on this one. We have the admin bar removed in the other modes/themes and I'm not immediately sure what differences in the legacy theme might account for this.

pierlon commented on assets/src/setup/pages/save/index.js:
```diff
@@ -31,26 +114,85 @@ export function Save() {
+				) }
+
+				{
+					'reader' === themeSupport && (
```
Can the buttons be placed to the left under the description? I don't think most users would be able to see them unless they were to scroll down first:

![](https://user-images.githubusercontent.com/16200219/86258786-bb558580-bbaa-11ea-89bd-af27d25982ad.gif)

amedina replied:
That may be good. Could we try it and see the difference?

amedina replied:
Also, why do we need a difference between the button for "`Visit in Customizer`" and "`Browse AMP Version`"?

Making them equal will provide symmetry.

johnwatkins0 replied:
Here's how that looks. Shall I push it @jwold @amedina ?

<img width="1381" alt="Screen Shot 2020-07-01 at 10 47 16 AM" src="https://user-images.githubusercontent.com/9020968/86264531-42ddcd00-bb88-11ea-90b9-3077d02d1dae.png">

johnwatkins0 replied:
Updated in 48653bbd1

johnwatkins0 committed:
Wizard done screen: add notices to indicate theme was downloaded or failed to downlaod

johnwatkins0 committed:
Remove debug code

johnwatkins0 committed:
Clarify 'Live view of your site' info for transitional/reader

pierlon commented:
Following up from review with 2 questions:

- Should we persist the local "developer tools" and "template mode" options? Otherwise the user has to re-select their chosen options each time if they were to navigate back to the page:

    ![](https://user-images.githubusercontent.com/16200219/86261313-eab9c180-bbad-11ea-9f98-76c1598eff57.gif)

- Should the template notice body be clickable?

    ![image](https://user-images.githubusercontent.com/16200219/86261522-31a7b700-bbae-11ea-84e3-9b4097263c80.png)

johnwatkins0 committed:
Setup wizard: make navbar have solid background with shadow

johnwatkins0 commented:
> > @jwold What do you think about @schlessera's suggestion on the bottom nav bar two comments up?
> 
> Let's go with @schlessera's change. I think it will be an improvement. cc @westonruter and @amedina

Done in edd43cb63

johnwatkins0 committed:
Setup wizard: remove arrow from 'Finish' button

johnwatkins0 committed:
Setup wizard: ensure nav is centered at wide widths

johnwatkins0 committed:
Move WP dependency to correct import section

johnwatkins0 committed:
Setup wizard: disable finish button on final screen while saving

johnwatkins0 committed:
Setup wizard: allow welcome screen to show while options are loading

johnwatkins0 committed:
Make panel buttons on template modes selection screen non-clickable

johnwatkins0 committed:
Update jest snapshot

johnwatkins0 commented:
> Following up from review with 2 questions:
> 
> * Should we persist the local "developer tools" and "template mode" options? Otherwise the user has to re-select their chosen options each time if they were to navigate back to the page:

Yes, this is a little annoying when doing QA, but it was agreed in one of our recent calls that we want to have user consciously make these choices when re-running the wizard. 

> * Should the template notice body be clickable?


Updated in 4139e4acd . In 1.7 these buttons will open panels; for now I just added `pointer-events: none` to make them nonclickable.

westonruter commented:
> Yes, this is a little annoying when doing QA, but it was agreed in one of our recent calls that we want to have user consciously make these choices when re-running the wizard.

The technical user question should persist once the wizard is completed, no?

@johnwatkins0 Also, in regards to the template mode selection, this brings up something I wanted to raise. The template mode should only be unselected (with the "Previously selected" label) if the user _changed_ the value for the technical user question. If I go through the wizard the first time, and I say I'm technical and set the mode to Transitional. The next time I go through the wizard, then the technical option should be pre-selected and if I leave that alone then the Transitional mode will be pre-selected on the next screen as well. _But_ if I change the response to the technical question then at that point no mode should be pre-selected.

johnwatkins0 committed:
Update logic behind when options are selected and unselected

johnwatkins0 committed:
Update test

johnwatkins0 committed:
Expand e2e tests

johnwatkins0 committed:
Update e2e tests

johnwatkins0 committed:
Update e2e tests

johnwatkins0 committed:
Lint JS fix

johnwatkins0 committed:
Add e2e tests for summary screen

johnwatkins0 committed:
Move e2e tsts into different directory to allow specificying test sequence

johnwatkins0 committed:
Fix persistence of mobile redirect

johnwatkins0 commented:
@westonruter @pierlon This is ready for review again. This PR has become a catch-all for everything we've discussed for the last couple days, and this afternoon while waiting for builds to run I refactored the e2e tests (to allow us to be able to sequence the tests -- which helps with testing persistent options). Sorry about the extra diff. Let me know if you see any issues.

westonruter commented:
@johnwatkins0 I went through the wizard after having saved previously with a Reader theme.

1.  The technical question is pre-selected to be non-technical. ✅
2. The Reader mode is selected by default. ✅ 
3. Clicking next (without changing the mode) then takes me _immediately_ to the Summary screen. I should have been able to go to the theme selection. ❌ 
4. The mobile theme screenshot is not populated. ❌ 
5. Clicking Previous takes me to Theme selection but the screen is empty. ❌ 

![image](https://user-images.githubusercontent.com/134745/86302450-99b1c980-bbbd-11ea-902b-fc9a9d0cb9ae.png)

![image](https://user-images.githubusercontent.com/134745/86302489-b6e69800-bbbd-11ea-9a3e-b812c3c35dc7.png)

To work around this issue I have to make a change to the mode and then switch back to Reader, and then I am taken to the theme selection and see the screenshot as expected on the summary screen.

johnwatkins0 committed:
Pull current reader theme from  instead of

westonruter committed:
Fix hiding the admin bar in legacy reader mode

johnwatkins0 commented:
Thanks, @westonruter . 7ab0495e8cc11798adeabb73644f669c3fac13a5 should fix that. I had to make some changes to the way options are held in the application state, so that now we have `originalOptions` (what's received from the endpoint), `updates` (changed values) and `editedOptions` ( `{ ...originalOptions, ...updates } )`, and it looks I missed switching `updates` to `editedOptions` in a couple places where the latter should be used.

westonruter commented:
@johnwatkins0 When clicking the Finish button I'm getting a JS error:

![image](https://user-images.githubusercontent.com/134745/86303233-19409800-bbc0-11ea-991d-1cc8c3bcda58.png)

westonruter commented:
`activePageIndex` is `5` so and there is no `activePageIndex + 1` page:

![image](https://user-images.githubusercontent.com/134745/86303289-42612880-bbc0-11ea-9f83-f333a0f6439a.png)

johnwatkins0 committed:
Skip 'next' button callback on last page when button is a link

johnwatkins0 committed:
Merge branch 'feature/4707-done-screen' of https://github.com/ampproject/amp-wp into feature/4707-done-screen

johnwatkins0 commented:
> @johnwatkins0 When clicking the Finish button I'm getting a JS error:
> 
> ![image](https://user-images.githubusercontent.com/134745/86303233-19409800-bbc0-11ea-991d-1cc8c3bcda58.png)

a56bd978e should fix that.

westonruter commented on assets/src/setup/pages/save/index.js:
```diff
@@ -3,25 +3,136 @@
+					className="done__preview-iframe"
+					src={ previewPermalink }
+					title={ __( 'Site preview', 'amp' ) }
+					name="amp-wizard-completion-preview"
```
Minor suggestion that I'm not totally sold on myself: since the admin bar gets hidden at `DOMContentLoaded` there is a flash of the admin bar. This could be avoided if we hide the `iframe` until the page is loaded.

```suggestion
					name="amp-wizard-completion-preview"
					style={ { opacity: 0 } }
					onLoad={ ( event ) => { event.target.style.opacity = '1'; } }
```

There's probably a more elegant way to do this however. (Including showing a spinner initially.) 

Not important.

johnwatkins0 replied:
Done in 9b707af3265b3f67067f45029f5030ef8d2c8221

westonruter commented:
@johnwatkins0 When on the Done screen, the Close button and the Finish button do the same thing:

![image](https://user-images.githubusercontent.com/134745/86304233-1c895300-bbc3-11ea-88a9-3e6c59e04484.png)

Should the close button be removed on the Done screen then? There's no sense to “Exit” the wizard when on the Done screen as the icon implies.

Also, I think the Close button should intentionally do `history.back()` to take the user to the screen they were just on (or else pass the return URL via a `return` query param similar to how the Customizer implements the Close button there).

**Update:** Filed as https://github.com/ampproject/amp-wp/issues/4974.

westonruter commented on assets/src/setup/components/navigation-context-provider.js:
```diff
@@ -24,19 +23,16 @@ export const Navigation = createContext();
-	// Initialize page from URL `amp-setup-screen` parameter. If not set, current page is 0.
-	// This is primarily for testing.
-	const [ activePageIndex, setActivePageIndex ] = useState( () => {
-		const index = pages.findIndex( ( { slug } ) => slug === getQueryArg( global.location.href, 'amp-setup-screen' ) );
```
Aside: I miss this query var 😄  It was useful to be on a screen and reload after changes to avoid having to tap Next Next Next.

johnwatkins0 replied:
Should I put it back? I thought I was probably the only person using it.

westonruter replied:
Yes, please do. It's nice for development, no?

pierlon commented:
Should the text within the information panel on the template mode screen be selectable?

![](https://user-images.githubusercontent.com/16200219/86304523-9ee33880-bbfe-11ea-9d17-ca3230a5fad5.gif)

johnwatkins0 committed:
Add iframe onload handler and fix some tests

johnwatkins0 committed:
remove pointer-events none from nonfunctioning button

johnwatkins0 commented:
> Should the text within the information panel on the template mode screen be selectable?
> 
> ![](https://user-images.githubusercontent.com/16200219/86304523-9ee33880-bbfe-11ea-9d17-ca3230a5fad5.gif)

Updated in 8c320a6857170fe4a63d2105e428531fe91e4bd3. I think we can just let these be button-like because they will be genuine buttons in a few weeks.

westonruter commented on assets/src/setup/pages/summary/test/redirect-toggle.js:
```diff
@@ -19,6 +19,12 @@ jest.mock( '../../../components/options-context-provider' );
 
 let container;
 
+function waitASecond() {
+	return new Promise( ( resolve ) => {
+		setTimeout( resolve, 25 );
+	} );
+}
```
```
/home/travis/build/ampproject/amp-wp/assets/src/setup/pages/summary/test/redirect-toggle.js
  22:10  error  'waitASecond' is defined but never used  no-unused-vars
```

johnwatkins0 replied:
Sorry, debug code. I've removed it.

johnwatkins0 committed:
Remove unused function

westonruter reviewed (approved):
Functionality checks out. I'll merge once @pierlon gives his +1 on technical specifics.

pierlon reviewed (approved):
Looks good, ship it!

schlessera commented:
>

schlessera commented:
> This may be because you still have "classic" set as the reader theme, which won't be accepted by the validate_callback now that we've switched that slug to legacy.

Won't other folks have this as well? We'll probably need to ensure whatever value is found is handled gracefully (by silently setting to the correct default value, for example?).

pierlon commented:
@schlessera I think he meant the [`reader_theme`](https://github.com/ampproject/amp-wp/blob/4139b2de3da1b036c5cc1c42bdebe2c4df07e2a1/includes/options/class-amp-options-manager.php#L34) option, which is not in production as yet.
